Mathlib/Algebra/Hom/NonUnitalAlg.lean

```diff
@@ -0,0 +1,448 @@
+algebra structures, this is the same as a not-necessarily-unital morphism of algebras. -/
+structure NonUnitalAlgHom [Monoid R] [NonUnitalNonAssocSemiring A] [DistribMulAction R A]
+  [NonUnitalNonAssocSemiring B] [DistribMulAction R B] extends A →+[R] B, A →ₙ* B
+#align non_unital_alg_hom NonUnitalAlgHom
+
+@[inherit_doc NonUnitalAlgHom]
+infixr:25 " →ₙₐ " => NonUnitalAlgHom _
+
+@[inherit_doc]
+notation:25 A " →ₙₐ[" R "] " B => NonUnitalAlgHom R A B
+
+attribute [nolint docBlame] NonUnitalAlgHom.toDistribMulActionHom
+attribute [nolint docBlame] NonUnitalAlgHom.toMulHom
+
+/-- `NonUnitalAlgHomClass F R A B` asserts `F` is a type of bundled algebra homomorphisms
+from `A` to `B`.  -/
+class NonUnitalAlgHomClass (F : Type _) (R : outParam (Type _)) (A : outParam (Type _))
+  (B : outParam (Type _)) [Monoid R] [NonUnitalNonAssocSemiring A] [NonUnitalNonAssocSemiring B]
+  [DistribMulAction R A] [DistribMulAction R B] extends DistribMulActionHomClass F R A B,
+  MulHomClass F A B
+#align non_unital_alg_hom_class NonUnitalAlgHomClass
+
+-- Porting note: commented out, not dangerous
+-- attribute [nolint dangerousInstance] NonUnitalAlgHomClass.toMulHomClass
+
+namespace NonUnitalAlgHomClass
+
+-- Porting note: Made following instance non-dangerous through [...] -> {_ : ...} replacement
+-- See note [lower instance priority]
+instance (priority := 100) toNonUnitalRingHomClass {F R A B : Type _}
+    {_ : Monoid R} {_ : NonUnitalNonAssocSemiring A} {_ : DistribMulAction R A}
+    {_ : NonUnitalNonAssocSemiring B} {_ : DistribMulAction R B}
+    [NonUnitalAlgHomClass F R A B] : NonUnitalRingHomClass F A B :=
+  { ‹NonUnitalAlgHomClass F R A B› with coe := (⇑) }
+#align non_unital_alg_hom_class.non_unital_alg_hom_class.to_non_unital_ring_hom_class NonUnitalAlgHomClass.toNonUnitalRingHomClass
+
+variable {_ : Semiring R} {_ : NonUnitalNonAssocSemiring A} {_ : Module R A}
+  {_ : NonUnitalNonAssocSemiring B} {_ : Module R B}
+
+-- see Note [lower instance priority]
+instance (priority := 100) {F : Type _} [NonUnitalAlgHomClass F R A B] : LinearMapClass F R A B :=
+  { ‹NonUnitalAlgHomClass F R A B› with map_smulₛₗ := map_smul }
+
+instance {F R A B : Type _} [Monoid R] [NonUnitalNonAssocSemiring A] [DistribMulAction R A]
+    [NonUnitalNonAssocSemiring B] [DistribMulAction R B] [NonUnitalAlgHomClass F R A B] :
+    CoeTC F (A →ₙₐ[R] B)
+    where coe f :=
+    { (f : A →ₙ+* B) with
+      toFun := f
+      map_smul' := map_smul f }
+
+end NonUnitalAlgHomClass
+
+namespace NonUnitalAlgHom
+
+variable {R A B C} [Monoid R]
+
+variable [NonUnitalNonAssocSemiring A] [DistribMulAction R A]
+
+variable [NonUnitalNonAssocSemiring B] [DistribMulAction R B]
+
+variable [NonUnitalNonAssocSemiring C] [DistribMulAction R C]
+
+-- Porting note: Replaced with FunLike instance
+-- /-- see Note [function coercion] -/
+-- instance : CoeFun (A →ₙₐ[R] B) fun _ => A → B :=
+--   ⟨toFun⟩
+
```

Comment:
I don't understand Mathlib4 coercions properly yet. Are we supposed to add something like this here:
```suggestion

attribute [coe] toMulHom
attribute [coe] toDistribMulActionHom

```

Reply:
@eric-wieser you seem to understand coercions... what do you think?

Afaict, this is the only remaining issue. The rest lgtm.

Reply:
I suspect that doesn't actually help, because these projections aren't actually used to define coercions anywhere. The coercion is actually going via a `CoeTC F (mul_hom A B)` instance elsewhere that doesn't use the projection, but takes the whole thing apart and puts it back together again instead. I think right now that's implemented inline, so there is nothing to tag with a `coe`, and the coercion unfolds to a constructor application.

Reply:
Thanks, I think I'm beginning to understand.